lib/active_record/tasks/redshift_7_0_tasks.rb

```diff
@@ -0,0 +1,148 @@
+
+# Based on https://github.com/rails/rails/blob/v7.0.8.4/activerecord/lib/active_record/tasks/postgresql_database_tasks.rb
+
+require "tempfile"
+
+module ActiveRecord
+  module Tasks # :nodoc:
+    class RedshiftDatabaseTasks # :nodoc:
+      DEFAULT_ENCODING = ENV["CHARSET"] || "utf8"
+      ON_ERROR_STOP_1 = "ON_ERROR_STOP=1"
+      SQL_COMMENT_BEGIN = "--"
+
+      delegate :connection, :establish_connection, :clear_active_connections!,
+        to: ActiveRecord::Base
+
+      def self.using_database_configurations?
+        true
+      end
+
+      def initialize(db_config)
+        @db_config = db_config
+        @configuration_hash = db_config.configuration_hash
+      end
+
+      def create(master_established = false)
+        establish_master_connection unless master_established
+        connection.create_database(db_config.database, configuration_hash.merge(encoding: encoding))
+        establish_connection(db_config)
+      end
+
+      def drop
+        establish_master_connection
+        connection.drop_database(db_config.database)
+      end
+
+      def charset
+        connection.encoding
+      end
+
+      def collation
+        connection.collation
+      end
+
+      def purge
+        clear_active_connections!
+        drop
+        create true
+      end
+
+      def structure_dump(filename, extra_flags)
+        search_path = \
+          case ActiveRecord.dump_schemas
+          when :schema_search_path
+            configuration_hash[:schema_search_path]
+          when :all
+            nil
+          when String
+            ActiveRecord.dump_schemas
+          end
+
+        args = ["--schema-only", "--no-privileges", "--no-owner"]
+        args.concat(["--file", filename])
+
+        args.concat(Array(extra_flags)) if extra_flags
+
+        unless search_path.blank?
+          args += search_path.split(",").map do |part|
+            "--schema=#{part.strip}"
+          end
+        end
+
+        ignore_tables = ActiveRecord::SchemaDumper.ignore_tables
+        if ignore_tables.any?
+          args += ignore_tables.flat_map { |table| ["-T", table] }
+        end
+
+        args << db_config.database
+        run_cmd("pg_dump", args, "dumping")
+        remove_sql_header_comments(filename)
+        File.open(filename, "a") { |f| f << "SET search_path TO #{connection.schema_search_path};\n\n" }
+      end
+
+      def structure_load(filename, extra_flags)
+        args = ["--set", ON_ERROR_STOP_1, "--quiet", "--no-psqlrc", "--output", File::NULL, "--file", filename]
+        args.concat(Array(extra_flags)) if extra_flags
+        args << db_config.database
+        run_cmd("psql", args, "loading")
+      end
+
+      private
+        attr_reader :db_config, :configuration_hash
+
+        def encoding
+          configuration_hash[:encoding] || DEFAULT_ENCODING
+        end
+
+        def establish_master_connection
+          establish_connection configuration_hash.merge(
+            database: "redshift",
```

Comment:
Diff here compared to https://github.com/rails/rails/blob/v7.0.8.4/activerecord/lib/active_record/tasks/postgresql_database_tasks.rb#L98

Reply:
I don't understand why you are overriding the configured database from the database.yml file.

running db:create now breaks
example:
```
production:
  primary:
    <<: *defaults
    adapter: postgresql
    ...
  data_warehouse:
    adapter: redshift
    host: hostname
    port: 5439
    database: "analytics"
    username: username
    password:password
    encoding: utf8
```
So now it's looking for a config with the database "redshift"

Caused by:
PG::ConnectionBad: could not translate host name "redshift" to address: Name or service not known (PG::ConnectionBad)

Reply:
@MrNagoo the error message you shared suggests the hostname isn't right rather than the database name - is that the right error message?

Reply:
Ah, after some digging it seems that it is possible to [override the default DbName in CloudFormation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-redshift-cluster.html#cfn-redshift-cluster-dbname) (how our clusters were created) - we obviously chose "redshift", which I assumed was the default to match the PG default of "postgres" (as per linked diff above). 

It seems the default Redshift DBName is actually `dev` - so perhaps this should use `dev`, but somehow allow overriding via the DB config - maybe via an `initial_database` property?

Reply:
I was planning on leaving a PR around the public_schema_method with a guard statement. The default variable `connection_already_established = false` isn't possible to set while calling `rake db:create` In redshift, the database gets created with any name you prefer. Without a monkey patch I'd have to name my database "redshift" in the database.yml. 

```ruby
def public_schema_config
  if configuration_hash.transform_keys(&:to_sym).keys?(:database)
    return configuration_hash.merge(schema_search_path: "public")
  end
  
  configuration_hash.merge(database: "redshift", schema_search_path: "public")
end
```

I initially understood it as a keywords issue but I've noticed that in my monkey patch, I'm simply calling merge again on create_database and it's re-overriding the "redshift" merge that overrode "analytics"